File: 백준/Gold/2230. 수 고르기/수 고르기.js
const filePath = process.platform === 'linux' ? '/dev/stdin' : 'input.txt'
const input = require('fs')
  .readFileSync(filePath)
  .toString()
  .trim()
  .split('\n')

// M 이상이면서 가장 작은 차이를 출력
// 오름차순 정렬
// 투포인터

const [N, M] = input[0].split(' ').map(Number)
const arr = input.slice(1).map(Number)

arr.sort((a, b) => a - b)

let left = 0, right = 0
let minDiff = Infinity

while (right < N) {
  const diff = arr[right] - arr[left]

  if (diff >= M) {
    minDiff = Math.min(minDiff, diff)
    left++

  } else {
    right++
  }
}

console.log(minDiff)

File: 백준/Gold/2230. 수 고르기/수 고르기.py
import sys
input = sys.stdin.readline

def func(n, m, arr):
    left = 0
    right = 0
    answer = 2000000000

    while right < n:
        diff = arr[right] - arr[left]

        if diff < m:
            right += 1
        elif diff == m:
            return m
        elif diff > m:
            answer = min(diff, answer)
            left += 1

    return answer

n, m = map(int, input().split())
arr = [int(input()) for _ in range(n)]

arr.sort()

result = func(n, m, arr)

print(result)
